myia/parser.py

```diff
@@ -0,0 +1,434 @@
+correspond to functions, and jumping from one block to another is done by a
+tail call. Phi nodes become formal parameters of the function. The inputs to a
+phi node become actual parameters (arguments) passed at the call sites in the
+predecessor blocks.
+
+Note that variable names in this module closely follow the ones used in [1]_ in
+order to make it easy to compare the two algorithms.
+
+The parsing of Python functions is handled on 3 separate levels, mimicking the
+way that Python resolves variable names (local, globals, builtins):
+
+Global environment
+  There is usually one environment per Python process, although multiple can be
+  instantiated. Within a single environment, the same Python objects will
+  correspond to the same Myia object e.g. if two functions use the same
+  subroutine and are parsed, this subroutine will only be parsed once.
+
+Parser
+  There is one parser per user-defined function. A parser is responsible for
+  e.g. resolving variable names using the function's global scope and managing
+  the parsing process.
+
+Block
+  A single basic block exists for each block of code in a user-defined
+  function. Basic blocks are responsible for resolving variable names locally
+  during parsing and converting Python's imperative control flow structures to
+  a functional representation.
+
+.. [1] Braun, M., Buchwald, S. and Zwinkau, A., 2011. Firm-A graph-based
+   intermediate representation. KIT, Fakultät für Informatik.
+.. [2] Appel, A.W., 1998. SSA is functional programming. ACM SIGPLAN Notices,
+   33(4), pp.17-20.
+
+"""
+import ast
+import builtins
+import inspect
+import textwrap
+from types import FunctionType
+from typing import Any, Dict, List, Tuple, Type
+
+from myia.anf_ir import ANFNode, Parameter, Apply, Graph, Constant
+
+
+class Environment:
+    """Environment to parse a function in.
+
+    An environment consists of a mapping from Python objects to nodes. If a
+    Python object is missing from this mapping, the environment will try to
+    convert it into a Myia object e.g. create a Myia constant or parse a Python
+    function into a myia function.
+
+    The environment is also in charge of resolving variable names that could
+    not be resolved in the function's local and global namespace i.e. for
+    built-ins and for undefined variable names.
+
+    Lastly, the environment has a mapping from AST nodes to Myia nodes that the
+    parser uses e.g. for mapping binary operators to the correct primitives.
+
+    Functions parsed in different environments will share no nodes in common.
+    Functions parsed in the same environment will use the same Myia objects for
+    the same Python objects.
+
+    """
+
+    def __init__(self, object_map: Dict[Any, ANFNode],
+                 ast_map: Dict[Type[ast.AST], ANFNode]) -> None:
+        """Construct an environment."""
+        self.object_map = object_map
+        self.ast_map = ast_map
```

Comment:
Should these be different dictionaries? I mean, there's no chance of overlapping.

Reply:
Theoretically, there is:
```python
x = ast.If
def f():
    return x  # Returns the `cond` Myia primitive instead of the Python object
```
Right now we don't support arbitrary Python objects, but that could be a possibility in the future (kind of like Cython and Numba, we could just switch back and forth to the interpreter to handle Python code we can't handle).

More so than this theoretical overlap though, I prefer keeping them separate for typing/logic/clarity reasons.

Reply:
Fair enough.